### ISSUE TITLE
pkg/util/flock.Acquire() doesn't guarantee mutual exclusion

### ISSUE THREAD

**Issue by yifan-gu** [kubernetes/kubernetes#42929]
**BUG REPORT**:

**Kubernetes version latest master**:

**Environment**:
- **Ubuntu 16.04 LTS, Linux 4.2.0-42-generic**:

**What happened**:
The `flock.Acquire()` doesn't provide mutual exclusion. Two processes can both grab the same flock at the same time.

**What you expected to happen**:
Only one process should be able to grab the flock.

**How to reproduce it** (as minimally and precisely as possible):

```shell
cat testflock.go
package main

import (
	"fmt"
	"time"

	"k8s.io/kubernetes/pkg/util/flock"
)

func main() {
	if err := flock.Acquire("/tmp/foo.lock"); err != nil {
		fmt.Println("error:", err)
		return
	}

	fmt.Println("acquired")

	time.Sleep(time.Hour)
}

go build testflock.go; ./testflock
acquired

# open another terminal and run
./testflock
acquired
```

**Anything else we need to know**:

Although the flock doesn't work, the [watchForLockfileContention()](https://github.com/kubernetes/kubernetes/blob/v1.7.0-alpha.0/cmd/kubelet/app/server.go#L324) does work. Since we _usually_ specify `--exit-on-lock-contention` together with `--lock-file`, so the 1st process will exit anyway.
However we still need to fix the flock..

**Comment by yifan-gu** [comment]
cc @mikedanese

**Comment by yifan-gu** [comment]
Sent a PR to fix this (#42935)
